File: julia/2020/03_toboggan_trajectory/aoc202003.jl
# Advent of Code 2020, day 3: Toboggan Trajectory

module AOC202003

using Pipe

# Parse input
function parse_data(puzzle_input)
    @pipe (
        puzzle_input
        |> split
        .|> map(==('#'), collect(_))
        |> hcat(_...)
        |> permutedims
    )
end


# Solve part 1
function part1(slope)
    slope |> count_trees
end


# Solve part 2
function part2(slope)
    [
        count_trees(slope, right=1, down=1),
        count_trees(slope, right=3, down=1),
        count_trees(slope, right=5, down=1),
        count_trees(slope, right=7, down=1),
        count_trees(slope, right=1, down=2),
    ] |> prod
end


# Count the number of trees following the given direction through the slope
function count_trees(slope; right=3, down=1)
    max_y, max_x = size(slope)
    idx_ys = range(1, step=down, stop=max_y)
    idx_xs = @pipe range(1, step=right, length=length(idx_ys)) .|> mod1(_, max_x)

    count(zip(idx_ys, idx_xs)) do (idx_y, idx_x)
        slope[idx_y, idx_x]
    end
end


# Solve the puzzle for the given input
function solve(puzzle_input)
    data = puzzle_input |> parse_data
    part1(data), part2(data)
end


# Solve the problem for one file
function solve_file(file_path)
    "\n$(file_path)" |> println
    open(file_path) do fid
        read(fid, String) |> strip .|> solve
    end
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module

File: python/2020/03_toboggan_trajectory/aoc202003.py
"""AoC 3, 2020: Toboggan Trajectory"""

# Standard library imports
import math
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input"""
    return [[c == "#" for c in line] for line in puzzle_input.split("\n")]


def part1(data):
    """Solve part 1"""
    return sum(trees(data))


def part2(data):
    """Solve part 2"""
    return math.prod(
        [
            sum(trees(data, right=1, down=1)),
            sum(trees(data, right=3, down=1)),
            sum(trees(data, right=5, down=1)),
            sum(trees(data, right=7, down=1)),
            sum(trees(data, right=1, down=2)),
        ]
    )


def trees(slope, right=3, down=1):
    """Yield trees along the slope in the given direction

    ## Examples:

        #.#.
        #..#
        .#..
        ..#.
        ##..

    >>> slope = [[1, 0, 1, 0], [1, 0, 0, 1], [0, 1, 0, 0], [0, 0, 1, 0], [1, 1, 0, 0]]
    >>> list(trees(slope, right=2, down=1))
    [1, 0, 0, 1, 1]
    """
    max_y, max_x = len(slope), len(slope[0])

    idx_x = 0
    for idx_y in range(0, max_y, down):
        yield slope[idx_y][idx_x]
        idx_x = (idx_x + right) % max_x


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
